File: ios/BibleApp/Services/BibleService.swift
import Foundation

/// Registry of TranslationProviders. Routes chapter/search calls by
/// translationID. Add new providers (API.Bible, on-device SQLite, etc.)
/// by registering them here.
final class BibleService {
    static let shared = BibleService()

    private let providers: [String: TranslationProvider]
    private let order: [String]

    private init() {
        let all: [TranslationProvider] = [
            BundledJSONProvider(
                translation: Translation(id: "KJV",
                                         name: "King James Version",
                                         publisher: "Public domain (1611)"),
                resourceName: "KJV"
            ),
            BundledJSONProvider(
                translation: Translation(id: "BBE",
                                         name: "Bible in Basic English",
                                         publisher: "Public domain (1949)"),
                resourceName: "BBE"
            ),
            ESVProvider(),
            BollsProvider(
                translation: Translation(id: "NIV",
                                         name: "New International Version",
                                         publisher: "Biblica / Zondervan (via bolls.life)"),
                remoteCode: "NIV"
            ),
            BollsProvider(
                translation: Translation(id: "NKJV",
                                         name: "New King James Version",
                                         publisher: "Thomas Nelson (via bolls.life)"),
                remoteCode: "NKJV"
            ),
            BundledJSONProvider(
                translation: Translation(id: "TAM",
                                         name: "Tamil Bible (TOV)",
                                         publisher: "Bible Society of India — Tamil Old Version"),
                resourceName: "TAM"
            ),
            BundledJSONProvider(
                translation: Translation(id: "TEL",
                                         name: "Telugu Bible (TOV)",
                                         publisher: "Bible Society of India — Telugu Old Version"),
                resourceName: "TEL"
            ),
        ]
        var dict: [String: TranslationProvider] = [:]
        for p in all { dict[p.translation.id] = p }
        self.providers = dict
        self.order = all.map { $0.translation.id }
    }

    /// Every translation we know about, in display order. UI can use
    /// `provider(for:).isAvailable` to decide whether to gate selection.
    var availableTranslations: [Translation] {
        order.compactMap { providers[$0]?.translation }
    }

    func provider(for translationID: String) -> TranslationProvider? {
        providers[translationID]
    }

    func chapter(translationID: String, bookID: String, chapter: Int) async throws -> [Verse] {
        guard let p = providers[translationID] else {
            throw BibleServiceError.translationNotBundled(translationID)
        }
        return try await p.chapter(bookID: bookID, chapter: chapter)
    }

    func search(translationID: String, query: String) async throws -> [Verse] {
        guard let p = providers[translationID] else {
            throw BibleServiceError.translationNotBundled(translationID)
        }
        return try await p.search(query: query)
    }
}

enum BibleServiceError: LocalizedError {
    case translationNotBundled(String)
    case bookNotFound(String)
    case chapterOutOfRange(String, Int)

    var errorDescription: String? {
        switch self {
        case .translationNotBundled(let id): "Translation \(id) is not configured."
        case .bookNotFound(let id): "Unknown book: \(id)."
        case .chapterOutOfRange(let id, let c): "\(id) has no chapter \(c)."
        }
    }
}

File: web/src/lib/bibleService.ts
import "server-only";
import { BundledJSONProvider } from "./providers/bundled";
import { ESVProvider } from "./providers/esv";
import { BollsProvider } from "./providers/bolls";
import type { TranslationProvider } from "./providers/types";
import type { Translation, Verse } from "./types";

const PROVIDERS: TranslationProvider[] = [
  new BundledJSONProvider(
    { id: "KJV", name: "King James Version", publisher: "Public domain (1611)" },
    "KJV"
  ),
  new BundledJSONProvider(
    { id: "BBE", name: "Bible in Basic English", publisher: "Public domain (1949)" },
    "BBE"
  ),
  new ESVProvider(),
  new BollsProvider(
    { id: "NIV", name: "New International Version", publisher: "Biblica / Zondervan (via bolls.life)" },
    "NIV"
  ),
  new BollsProvider(
    { id: "NKJV", name: "New King James Version", publisher: "Thomas Nelson (via bolls.life)" },
    "NKJV"
  ),
  new BundledJSONProvider(
    { id: "TAM", name: "Tamil Bible (TOV)", publisher: "Bible Society of India — Tamil Old Version" },
    "TAM"
  ),
  new BundledJSONProvider(
    { id: "TEL", name: "Telugu Bible (TOV)", publisher: "Bible Society of India — Telugu Old Version" },
    "TEL"
  ),
];

const BY_ID: Record<string, TranslationProvider> =
  Object.fromEntries(PROVIDERS.map(p => [p.translation.id, p]));

export interface TranslationSummary extends Translation {
  available: boolean;
  reason: string | null;
}

export function listTranslations(): TranslationSummary[] {
  return PROVIDERS.map(p => ({
    ...p.translation,
    available: p.isAvailable(),
    reason: p.unavailabilityReason(),
  }));
}

export function getProvider(translationID: string): TranslationProvider | undefined {
  return BY_ID[translationID];
}

export async function chapter(
  translationID: string, bookID: string, chapter: number,
): Promise<Verse[]> {
  const provider = BY_ID[translationID];
  if (!provider) throw new Error(`Translation ${translationID} is not configured`);
  return provider.chapter(bookID, chapter);
}

export async function search(translationID: string, query: string): Promise<Verse[]> {
  const provider = BY_ID[translationID];
  if (!provider) throw new Error(`Translation ${translationID} is not configured`);
  return provider.search(query);
}
